astropy/table/table.py

```diff
@@ -732,6 +732,17 @@ def __init__(
         # a list of dict. If data are not list of dict then this is None.
         names_from_list_of_dict = None
 
+        # Treat empty rows same as None case.
+        if isinstance(rows, (list, tuple, np.ndarray)):
```

Comment:
@neutrinoceros @taldcroft , looks uglier but should work for all the cases you brought up thus far?

Reply:
I feel like this check might be too intrusive and possibly could introduce performance regression in some cases. Ideas welcome!